src/Types.elm

```diff
@@ -2,18 +2,76 @@ module Types exposing (..)
 
 type Tree a = Node a (List (Tree a))
 
-type Token = TTSC TokTSC | TokAssign | TokLParen | TokRParen | TokHole | TokVar String
-             | TokConstInt Int | TokConstBool Bool | TokInvalid | TokBackSlash | TokArrow | TokEnd
+type Token = TTSC TokTSC | TokAssign | TokLParen | TokRParen | TokHole |
+              TokVar String | TokHasType | TokConstInt Int | TokConstBool Bool |
+              TokInvalid | TokBackSlash | TokArrow | TokEnd | TokTypeName String
 
 type TokTSC = TTSCInt TokTSCInt | TTSCBool TokTSCBool
 type TokTSCInt = TokPlus | TokMinus | TokTimes
 type TokTSCBool = TokEq | TokAnd | TokOr
 
 type Const = CBool Bool | CInt Int
 type Term = CTerm Const | VTerm String | Plus Term Term | Minus Term Term | Times Term Term
-            | Eq Term Term | And Term Term | Or Term Term | Lam Term Term | App Term Term
+            | Eq Term Term | And Term Term | Or Term Term | Lam String Term | App Term Term
             | MissingInt | MissingBool | Missing | EmptyTree
 
+-- V(alue)Type is a type that a TreeAssembly term can evaluate to
+{-
+  TAny may be used when the type can be either int or bool (discuss)
+-}
+type VType = TBool | TInt | TFun VType VType
+
 type alias Var =
   { name: String
-  , term: Term }
+  , term: Term
+  , vtype: VType}
+
+{-
+  Returns the type signature of basic language constructs (only used for binary operators)
+-}
+getTypeSignature : Term -> Maybe VType
+getTypeSignature t =
+  case t of
+    CTerm (CBool b) -> Just TBool
+    CTerm (CInt i) -> Just TInt
+    Plus x y -> Just (TFun TInt (TFun TInt TInt))
+    Minus x y -> Just (TFun TInt (TFun TInt TInt))
+    Times x y -> Just (TFun TInt (TFun TInt TInt))
+    Eq x y -> Just (TFun TBool (TFun TInt TInt))
```

Comment:
Good changes here!  I think Eq should be `Just (TFun TInt (TFun TInt TBool))`.  (Like in Haskell or Elm we'd write the signature as `Int -> Int -> Bool`.  If you take out `TFun` in our definition you can kinda see the Haskell/Elm definition.)

Reply:
Fixed!